### DISCUSSION TITLE
Does the core library have plans to support reactive programming?

### DISCUSSION THREAD

**Discussion by jeffwcx** [moonbitlang/core#336] [Announcements]
Reactive programming is quite useful in the field of GUI development. Is there consideration for adding a class like signal?

**Comment by hackwaly** [comment]
Good idea. I have a WIP version of signal, but not applied `@preact/signal`'s optimization.

**Comment by jeffwcx** [comment]
Maybe we should propose an RFC to discuss the API

**Comment by hackwaly** [comment]
Nice

**Comment by snnsnn** [comment]
Reactive programming is indeed quite powerful for GUI development, especially when it comes to managing state and reacting to user interactions in a clean, declarative way.

That said, it’s still a bit early to include a reactive primitive like signal in the core library. There isn’t yet a widely agreed-upon approach or standard for reactive primitives across the broader ecosystem. Different frameworks and communities have their own implementations (e.g., Solid’s createSignal, S.js, MobX, etc.), each with their own trade-offs in terms of ergonomics, performance, and complexity.

It makes sense to let libraries experiment and evolve naturally—let the best implementations emerge through usage and community feedback. Over time, the most practical and well-designed patterns tend to stand out. Once there's clearer consensus, that would be a more solid foundation for any potential inclusion in the core.
